### PR TITLE
Feature/generic gpu optimization

### PR DESCRIPTION
**Pull request by gpinkert**
## Summary

Adds a fourth Hyperloom skill, **`generic-gpu-optimization`**, that lets the agent optimize **arbitrary GPU codebases** — not just SGLang/vLLM inference servers (`inference-optimization`) or Primus/Megatron training (`training-optimization`). The skill auto-detects the project shape and runs the same DFS Think → Try → Measure → Decide loop against detected build / bench / test commands.

The existing four skills are untouched; this is purely additive.

## Motivation

Today, pointing Hyperloom at a project that isn't an inference server or a Primus/Megatron training stack means hand-writing a custom skill: rewriting `setup.md`, `classify.md`, `run_baseline.sh`, swapping `torch.profiler` for `rocprofv3`, etc. That's a lot of boilerplate for what is fundamentally the same DFS loop with three abstract operations: *build*, *measure*, *test*.

This skill encodes those three operations as a heuristic detector and a set of dispatchers, so that the agent can be invoked against e.g. a HIP/CMake library, a standalone PyTorch script, or a Triton kernel collection without skill-authoring overhead.

### Bootstrap any repo in one command

```bash
# from the Hyperloom checkout
./scripts/optimize_repo.sh /path/to/your/gpu/repo
```

This:
- copies `.cursor/skills/generic-gpu-optimization/` into the target
- copies `.cursor/mcp.json` (GEAK + OOB MCPs)
- copies `.env.template`
- runs the heuristic detector and prints what was found

### Run from Cursor

After setting `AK_YOUR_API_KEY` in `<target>/.env` and enabling the GEAK + OOB MCP toggles:

```
@.cursor/skills/generic-gpu-optimization/SKILL.md
Optimize this repo for MI355X.
```

The agent runs **setup → detect → build → baseline → correctness → profile → DFS loop (env-vars / compile-flags / GEAK kernel-opt) → report**. Every kept change is gated against the project's own test suite (ctest / pytest); regressions and correctness failures auto-revert.

### Override anything

If the detector got something wrong, override in the prompt:

```
@.cursor/skills/generic-gpu-optimization/SKILL.md
Optimize this repo for MI300X.

BENCH_COMMAND: ./build/bench/MATRIX_BENCH --benchmark_filter=select_k --benchmark_format=json
TEST_COMMAND:  ctest -R MATRIX_TEST --output-on-failure
TIME_BUDGET_MIN: 60
```

## What gets detected

| Project signal | Classified as | Default build | Default bench | Default tests |
|---|---|---|---|---|
| `CMakeLists.txt` + `*.hip` + `benchmark::benchmark` link | `hip-cmake-bench` | `cmake --build` (or `build.sh` if present) | first executable matching `*bench*` in build dir | `ctest` (located in actual `CTestTestfile.cmake` dir) |
| `pyproject.toml` + `import torch` + `bench*.py` | `pytorch-script` | `pip install -e .` | `python <entrypoint>` | `pytest -x` |
| `@triton.jit` + `bench*.py` or `benchmarks/` | `triton-collection` | `pip install -e .` | `python bench.py` | `pytest -x` |
| `Makefile` + `*.hip` | `hpc-app` | `make -j$(nproc)` | (user override required) | (user override) |
| (none of the above) | `unknown` | (user override required) | (user override) | (user override) |

The detector skips vendored trees (`thirdparty/`, `_deps/`, `.venv/`, `site-packages/`, `build*/`) so it doesn't mis-classify based on code you don't own.

## Action repertoire

| Action | Cost | Risk | Typical gain |
|---|---|---|---|
| `env-vars` | seconds (no rebuild) | low | 1–8 % |
| `compile-flags` | minutes (rebuild) | medium (`-ffast-math` can break things) | 2–15 % |
                                                                                            | `kernel-opt` (GEAK on hot HIP/Triton kernel) | 10–30 min/kernel | medium (output validated by tests) | 5–50 % |
| `patch` (manual code change) | minutes | medium-high | varies |
| `torch.compile` (`pytorch-script` only) | minutes | low | 5–15 % |

Initial priors are project-class specific. Re-scoring rules mirror `training-optimization` (success boosts similar actions ×1.5, failure penalizes ×0.5).

## Correctness gating

Every kept change must pass the project's own tests. The skill never invents tests:

- `CTestTestfile.cmake` present → `ctest --output-on-failure`
- `tests/test_*.py` present → `pytest -x`
- neither → capture baseline bench output as golden, compare numerical fields
- nothing parseable → warn user, accept on bench exit-code only

Mandatory gate. A 10× speedup that fails tests is reverted, recorded as INVALID, and the agent moves on.

## Profiling

| Stack | Profiler | Output schema |
|---|---|---|
| `pytorch-script` / `triton-collection` | `torch.profiler` (Chrome trace) | `traceEvents[].cat == "kernel"` |
| `hip-cmake-bench` / `hpc-app` | `rocprofv3` (preferred), `rocprof v2` fallback | `rocprofiler-sdk-tool[].buffer_records.kernel_dispatch[]` |

Both are normalized by `scripts/analyze_profile.py` into a top-N kernel breakdown, so `kernel-opt.md` is profiler-agnostic.

## Verified end-to-end on a real codebase

Smoke-tested against [hipRAFT](https://github.com/AMD-AIOSS/hipRaft) on MI210 (gfx90a) / ROCm 7.2.0:

| Stage | Result |
|---|---|
| bootstrap | Skill, mcp.json, .env.template copied; detector summary printed |
| setup | MI210 / gfx90a / 8 GPUs / ROCm 7.2.0 detected |
| detect | Classified `hip-cmake-bench`, found `build.sh` + `cpp/build/tests/CTestTestfile.cmake`, `KERNEL_LANGS="hip"` |
| correctness | `gtests/UTILS_TEST --gtest_filter='*Bitonic*'` → 45/45 passed in 373 ms |
| baseline | `run_bench.sh` + `extract_metric` → 367 ms |
| profile | `rocprofv3` → 55 MB trace; `analyze_profile.py` → top kernel `raft::util::bitonic_kernel<128, double>` at **21.1 %** of GPU time |
| attempt cycle | 3 env-var attempts ran cleanly through bench → correctness → keep/revert; `kept_env.sh` accumulated wins |

The trace surfaced a real, hardware-grounded finding the skill is designed to expose:

> Every Capacity ≥ 16 instantiation of `bitonic_kernel` is spilling up to **2700 bytes/thread to global scratch**, with **zero AGPRs being used** despite gfx90a having 128 of them per wave. Top three proposed actions:
> 1. compile-flag `-mllvm -amdgpu-spill-vgpr-to-agpr=true`
> 2. patch `local_arr` from registers to LDS at `cpp/tests/util/bitonic_sort.cu:75`
> 3. add `__launch_bounds__(kMaxBlockSize, 2)` to double occupancy

(Negative findings: `GPU_MAX_HW_QUEUES=16` regresses by 27.8 % on this workload — saved as a KB entry candidate so the agent doesn't re-test it.)

## Files added

```
.cursor/skills/generic-gpu-optimization/
├── SKILL.md                 # DFS orchestrator (project-class agnostic)
├── README.md                # Usage guide + worked example
├── actions/                 # 11 playbooks
│   ├── setup.md             # GPU/ROCm detection, RESULT_DIR
│   ├── detect.md            # Auto-discover build/bench/test/kernel-langs
│   ├── build.md             # Dispatches to detected build_command
│   ├── baseline.md          # Runs bench, extracts metric, noise check
│   ├── correctness.md       # Runs detected tests; tests / golden-output / none modes
│   ├── profile.md           # Picks rocprofv3 or torch.profiler
│   ├── env-vars.md          # 11 ROCm runtime env-var candidates
│   ├── compile-flags.md     # CMake / HIP / Clang flag candidates
│   ├── kernel-opt.md        # GEAK MCP submission; works on any HIP/Triton kernel
│   ├── patch.md             # Generic code-patch action with revert protocol
│   └── report.md            # Final report renderer
├── scripts/
│   ├── common.sh            # extract_metric, kill_workload, etc.
│   ├── detect_project.sh    # Heuristic project classifier
│   ├── run_bench.sh         # Dispatcher
│   ├── run_correctness.sh   # Dispatcher
│   ├── profile_rocm.sh      # rocprofv3 wrapper (with rocprof v2 fallback)
│   ├── profile_torch.sh     # torch.profiler wrapper
│   └── analyze_profile.py   # Normalizes 3 profiler schemas → top-N breakdown
└── kb/entries.jsonl         # Empty seed (re-uses training-optimization KB schema)

scripts/optimize_repo.sh     # One-shot bootstrap into a target repo

docs/GENERIC_GPU_OPTIMIZATION.md   # Design + auto-detection rules + skill comparison

README.md                    # New "Quickstart — Any GPU Repo" section, table entry, repo-tree, Learn More link
```
## Commits

The branch is two commits:

- **`0585b48`** `feat: add generic-gpu-optimization skill for arbitrary GPU repos` — initial implementation
- **`f297896`** `fix: bugs surfaced by end-to-end smoke test on hipRAFT` — six real bugs caught during the live smoke run:
  1. `GPU_COUNT` was always wrong (counted lines, not devices) → fixed to `sort -u` on `GPU[N]` indexes
  2. `KERNEL_LANGS="triton"` false-positive (`head | true` exits 0 on empty) → fixed to `[ -n "$(...)" ]`; also added vendored-tree exclusions
  3. `TEST_COMMAND="cd build && ctest"` ignored real `cpp/build/tests/` location → finds actual `CTestTestfile.cmake` dir
  4. `BENCH_COMMAND` placeholder embedded `$BUILD_DIR` → broke `set -u` sourcing; split into `BENCH_HINT`
  5. `optimize_repo.sh --force` silently nested files (`cp -r src dst` when `dst` is a dir) → added `rm -rf` first
  6. `profile.md` analyzer used a guessed schema → rewrote to handle real rocprofv3 (`rocprofiler-sdk-tool` with `kernel_symbols` lookup), promoted to standalone `analyze_profile.py`

## Comparison with the existing skills

| Concern | inference-opt | training-opt | generic-gpu-opt |
|---|---|---|---|
| Server lifecycle | Yes (SGLang/vLLM) | No | No |
| YAML config parsing | No | Yes (Primus) | No |
| GBS gate | No | Yes (immutable global batch) | No (uses test suite) |
| Build system | n/a (pip) | n/a (pip) | Detected: cmake/pip/make |
| Profiler | torch.profiler + TraceLens | torch.profiler + TraceLens | torch.profiler OR rocprofv3 |
| GEAK kernel-opt | Yes | Yes | Yes |
| Initial priors | model-class table | model-class table | project-class table |
| Stopping criteria | identical | identical | identical |

Full design + auto-detection rules: [`docs/GENERIC_GPU_OPTIMIZATION.md`](docs/GENERIC_GPU_OPTIMIZATION.md). Skill README: [`.cursor/skills/generic-gpu-optimization/README.md`](.cursor/skills/generic-gpu-optimization/README.md).

## Untested

- `pytorch-script` path — no suitable non-inference PyTorch project on the test box
- `kernel-opt` action against the live GEAK MCP — needs Cursor + a real `AK_YOUR_API_KEY`
- `compile-flags` action through a full hipRAFT rebuild — would take ~10+ min; the action calls `build.md` which is exercised by `hip-cmake-bench` patterns but not yet end-to-end with a flag swap

### CODE REVIEW

**Comment by ZhengGong-amd** [comment]
Thanks for putting this together — generalizing the Hyperloom DFS loop beyond SGLang/vLLM and Primus/Megatron is a direction we've discussed internally, and the abstraction you landed on is clean and consistent with the existing skills.

After discussing with the team, we'd like to **hold off on merging this into `main` for now** and ask you to keep iterating on the feature branch. Two reasons, neither about the design itself:

1. **Maturity gap vs. the other skills in `main`.** The only end-to-end validation so far is the hipRAFT smoke test in `f297896`, which surfaced six real bugs in a single pass (GPU count, `KERNEL_LANGS` misdetection, hard-coded `ctest` path, non-sourceable `detected.env`, etc.). The other three skills each ship with published speedups and a written case study, and we'd like this one to clear a similar bar before it lands next to them.

2. **No measured results yet.** We noticed that `kb/entries.jsonl` is an empty seed, and there are no kept patches or numbers from a real optimization run. Without at least one worked example with each claimed backend, it's hard for downstream users to know when this skill is the right tool to reach for.

As you've already noted, `scripts/optimize_repo.sh` acts as an out-of-tree bootstrapper, so users who want to try it can do so directly from the feature branch without needing it in `main`. That gives you room to keep iterating without a compatibility commitment.

Really appreciate the work.
